File: Oil_Asian_opt_Additive_Bachelier/Python/utils/charFunction.py
import numpy as np
from scipy.special import gamma


def psi(u: np.ndarray, kappa: float, eta: float, alpha: float = 1 / 3) -> np.ndarray:
    if alpha == 0:
        return -1 / kappa * np.log(1 + kappa * u)
    else:
        return (
            (1 / kappa)
            * (1 - alpha)
            / alpha
            * (1 - (1 + kappa * u / (1 - alpha)) ** alpha)
        )


def charFunction(
    sigma: float, t: float, kappa: float, eta: float, alpha=1 / 3
) -> callable:
    def cf(u):
        u = np.asarray(u, dtype=np.complex128)
        arg1 = 1j * u * eta * sigma * np.sqrt(t) + 0.5 * (u**2) * sigma**2 * t
        return np.exp(psi(arg1, kappa, eta, alpha) + 1j * u * eta * sigma * np.sqrt(t))

    return cf


def beta_complex(
    z: np.complex128 | np.ndarray, w: np.complex128 | np.ndarray
) -> np.complex128 | np.ndarray:
    """
    Compute the beta function for complex arguments.
    :param z: Complex number or array of complex numbers.
    :return: Beta function value.
    """
    return gamma(z) * gamma(w) / gamma(z + w)


def charFunctionADL(b: float):
    def cf(u):
        u = np.asarray(u, dtype=np.complex128)
        return (1 - b) * beta_complex(1 + (1j * u - 1) * b, 1 - 1j * u * b)

    return cf

File: Oil_Asian_opt_Additive_Bachelier/utils/charFunction.m
function phi = charFunction(sigma,t, k, eta)
% Characteristic Function of the Additive Bachelier Model.
% The actual expression depends on the hyper parameter alpha. For typing
% simplicity, we have opted to not make alpha an actual function input,
% since it's always the same throughout the code and we didn't like how 
% MATLAB manages default arguments.
% THIS METHOD RETURNS A FUNCTION HANDLE
% INPUTS:
% sigma -> float, level of volatility at time t
% t -> float, time to maturity (year frac)
% k -> float, vol of vol
% eta -> float, skewness

% alpha = 1/2;
alpha = 1/3;

if alpha == 0
    % G is a Gamma distribution with parameters a and b
    psi = @(u) -1/k*log(1+k*u);
else
    % G is an Inverse Gaussian
    psi = @(u) 1/k*(1-alpha)/alpha*(1-(1+(u*k)./(1-alpha)).^alpha);
end
   
logphi = @(u) psi(1i*u*eta*sigma*sqrt(t)+u.^2/2*sigma.^2.*t)+ ...
              1i*u*eta*sigma.*sqrt(t);
phi = @(u) exp(logphi(u));

end